taichi/analysis/same_statements.cpp

```diff
@@ -97,22 +120,57 @@ class IRNodeComparator : public IRVisitor {
       same = false;
       return;
     }
+    auto other = other_node->as<Stmt>();
 
     // If two identical statements can have different values, return false.
-    if (check_same_value_ && !stmt->is_container_statement() &&
-        !stmt->common_statement_eliminable()) {
-      same = false;
-      return;
+    // TODO: two identical GlobalPtrStmts cannot have different values,
+    //  but GlobalPtrStmt::common_statement_eliminable() is false.
```

Comment:
Do you mean "when ... is false"?

Reply:
We want to know if two identical statements of the type same as `Stmt` can have different values, for example, two identical `RandStmt`s can have different values.

In most cases, this property is the same as `Stmt::common_statement_eliminable()`: two identical `RandStmt`s cannot be CSE-ed.

However, for global pointers, `GlobalPtrStmt::common_statement_eliminable()` is false because there is an `activate` field in `GlobalPtrStmt`s. But two identical `GlobalPtrStmt` definitely point to the same address, so in `same_value`, we do not want to set `same = false` here.